### ISSUE TITLE
Passing incorrect parameters to the CrudPanel methods should result in errors/exceptions, not unexpected behavior

### ISSUE THREAD

**Issue by tumf87** [Laravel-Backpack/CRUD#985]
While writing the CrudPanel unit tests I've found a number of methods in this situation:

- `addButton` method in the Buttons trait -> adding a button on an unknown position actually adds the button at the end of the stack
- `addButton` method in the Buttons trait -> adding a button on an unknown stack adds the button to the button list without any validation (there may be a use case for this in case users create their own stack)
- `getRelationFields` method in the Create trait -> getting relations fields from an unknown form actually returns the fields from the update form
- `getFields` method in the Read trait -> getting fields from an unknown form returns the create form fields
- `getFakeColumnsAsArray` method in the FakeColumns trait -> this calls the `getFields` method with the same unknown form, which in turn returns the create form fields
- `compactFakeFields` in the FakeFields trait -> this calls the `getFields` method with the same unknown form, which in turn returns the create form fields
- `addFields` in the Fields trait -> no validation is performed for a field of an unknown type (pre-defined or custom)

My view on this is that throwing exceptions in these cases would lead to fewer bugs and easier debugging than in the case of an unexpected behavior.

None of the fixes would imply breaking changes.

@tabacitu, what do you think? Can you think of some use cases that I may have missed?

**Comment by tabacitu** [comment]
Hmm... **My reasoning, in general, is that it's ok to throw exceptions, as long as you don't break people's existing apps, or prevent them from easily extending the software.** More strictness has to have a real UX benefit. I don't like strictness for strictness' sake.

So:
> 1. `addButton` method in the Buttons trait -> adding a button on an unknown position actually adds the button at the end of the stack

So you'd recommend restricting the ```position``` options to ```beginning``` and ```end```? Is that what you're suggesting? If so - I agree, I see no harm in doing this. I don't think anyone is using custom positions.  

> 2. `addButton` method in the Buttons trait -> adding a button on an unknown stack adds the button to the button list without any validation (there may be a use case for this in case users create their own stack)

Well... this allows people to easily define other stacks in their view, that would then "_just work_". 

If we were strict about the "stack" possible values here, they'd have to jump through hoops to do the same thing - they'd have to define that stack in the view AND somewhere else (most intuitive would be in the config file). But maybe they don't use that new stack everywhere, maybe they only use in in some CRUDs, with a custom ```list.blade.php``` and others not. So you'd also need an API for that, to specify a new button stack. The result would be that this minor _minor_ issue would imply creating a big breaking change (for some) and a huge inconvenience for the power user. All for what - just throwing an exception if the user has made a spelling mistake? Definitely not worth it, IMHO. Let the user just notice that the button isn't there. That should be obvious enough.

> 3. `getRelationFields` method in the Create trait -> getting relations fields from an unknown form actually returns the fields from the update form

Same as the above. Allows people to easily define custom operations that use fields. Don't know how many people do this, but I know of at least 1 case.

> 4. `getFields` method in the Read trait -> getting fields from an unknown form returns the create form fields

Same as above. But we should probably stick with one default result everywhere - either ```create``` or ```update```. Right now, ```create``` sounds like a more intuitive default to me. What do you think?

> 5. `getFakeColumnsAsArray` method in the FakeColumns trait -> this calls the `getFields` method with the same unknown form, which in turn returns the create form fields

Same as above. This would be automatically fixed by whatever solution we find for the above.

> 6. `compactFakeFields` in the FakeFields trait -> this calls the `getFields` method with the same unknown form, which in turn returns the create form fields

Same as above. This would be automatically fixed by whatever solution we find for the above.

> 7. `addFields` in the Fields trait -> no validation is performed for a field of an unknown type (pre-defined or custom)

Fields are picked up from the ```resources/views/vendor/backpack/crud/fields``` first (we check if they've been overwritten); if that doesn't exist, it looks in the same folder in the package. So there are no "known" and "unknown" field types. If the view is there, it should be available. I'm totally ok with throwing an exception if the view doesn't exist. But I bet Laravel already does that for us, when rendering the page. No use in checking when adding the field.

----

Bottom line, I think:
- (1) ok
- (2) not worth it
- (3) not worth it
- (4) not worth it; but we should probably choose one default, either create/update, and use it across the board;
- (5) non-issue; it just calls (4);
- (6) non-issue; it just calls (4);
- (7) I think Laravel already does this for us when rendering the page;

Either way, I'd say this is very low prio.

**Comment by tumf87** [comment]
@tabacitu, I'm not strict. It's all fake news, believe me!
Joking aside, we don't have different opinions on this, I'm all for excitability. It's at the core of every API. What I'm arguing about is the other core principle of an API: it being intuitive. A method should do what it says it does.

And now for some replies:

3. > Allows people to easily define custom operations that use fields.

Look at this code for the method `getRelationFields($form = 'create')`:

```php
if ($form == 'create') {
    $fields = $this->create_fields;
} else {
    $fields = $this->update_fields;
}
```
No matter what's supplied as the `$form`, if it's not 'create', this method will return the update fields. If a custom form is needed, this method won't just work, it will need to be overridden.
If we wanted to allow custom fields for any supplied value of `$form`, we will need to keep the fields in a key-value based array. This would be a breaking change. 

I'm not saying we should go this route yet, I'm saying we should change this method to return the create fields for `$form = 'create'`, the update fields for `$form = 'update'` and an exception otherwise, which will explicitly tell the user that he needs to override this method for custom forms.

4. Same exact issue with this code from the `getFields` method (and for the methods that call this method):

```php
switch (strtolower($form)) {
    case 'create':
    return $this->getCreateFields();
    break;

case 'update':
    return $this->getUpdateFields($id);
    break;

default:
    return $this->getCreateFields();
    break;
}
```

**Comment by lloy0076** [comment]
@tabacitu What's happening with this?

IMHO there are too many issues in this issue and I'm tempted to squash it and ask *someone* to reopen the issues individually if they really matter (honestly, they don't - this has been parked since last October).

**Comment by tabacitu** [comment]
Agreed. Let's focus on things that matter, we can reopen this whenever someone deems it important.
